protocols/relay/src/priv_client/handler.rs

```diff
@@ -140,8 +153,12 @@ pub struct Handler {
     /// eventually.
     alive_lend_out_substreams: FuturesUnordered<oneshot::Receiver<void::Void>>,
 
-    circuit_deny_futs:
-        HashMap<PeerId, BoxFuture<'static, Result<(), protocol::inbound_stop::UpgradeError>>>,
+    open_circuit_futs: futures_bounded::WorkerFutures<
+        (),
+        Result<inbound_stop::Circuit, inbound_stop::FatalUpgradeError>,
+    >,
+
+    circuit_deny_futs: HashMap<PeerId, BoxFuture<'static, Result<(), inbound_stop::UpgradeError>>>,
```

Comment:
How about making this a `futures_bounded` as well? Would implicitly enforce `MAX_NUMBER_DENYING_CIRCUIT`. In other words simplify the code below.

Reply:
Good point.
But there are a few questions:
1. We should add to `futures_bounded` ability to remember what it has keys, isn't?
2. What should we do when a timeout happens?

Reply:
> * We should add to `futures_bounded` ability to remember which it has keys, isn't?
>   Otherwise, it changes functionality, we could add the same key twice.

Overall, that would make sense but the problem is that `()` is always equal to `()`. Thus, if we enforce that `futures-bounded` can't have duplicate keys, then we can't use it with `()` as a key.

We'd have to special-case that in `futures-bounded` by creating separate `impl` blocks with functions that don't allow you to pass a worker ID and thus also don't enforce its uniqueness. I am not sure that is worth the effort?

> * What should we do when a timeout happens?

We can just log an error. At the moment, we don't have any timeout actually! This shows how important this PR is :)